### PR TITLE
Allow the gutter to show relative line numbers

### PR DESCRIPTION
**Pull request by Nadrieril**
Displays vim-like relative line numbers in gutter, with option to show true current line number instead of zero on current line
https://github.com/atom/atom/issues/1612

### CODE REVIEW

**Comment by 50Wliu** [comment]
You'll probably want to add tests for the new behavior in order for this to get merged.

**Comment by Nadrieril** [comment]
I'd like to do that indeed. Could you give me some pointers for how to run atom tests ? I can find it anywhere in the docs

File: src/gutter-component.coffee [hunk 1]
```diff
@@ -3,6 +3,7 @@ React = require 'react-atom-fork'
 {div} = require 'reactionary-atom-fork'
 {isEqual, isEqualForProperties, multiplyString, toArray} = require 'underscore-plus'
 Decoration = require './decoration'
+{$} = require './space-pen-extensions'
```
Comment (thedaniel) [inline src/gutter-component.coffee]: We're trying to remove jQuery dependencies wherever we can, this probably won't get merged with this require.

Comment (Nadrieril) [inline src/gutter-component.coffee]: It was indeed not necessary, I removed it

**Comment by 50Wliu** [comment]
@Nadrieril try [this document](https://github.com/atom/atom/blob/master/docs/writing-specs.md) (scroll down near the bottom to see how to run them).  Basically-`window:run-package-specs`.

**Comment by nathansobo** [comment]
@Nadrieril hey, so you should really be performing the logic for this in the `TextEditorPresenter` layer rather than in the view. Ideally it's implemented and tested mostly at the presenter layer and the view just has minor tweaks to consume it. We should probably replace "buffer row" with "display row" like you're doing now in the line number state objects in the presenter.

**Comment by Nadrieril** [comment]
I did not see so much changes coming.
I don't think we can replace entirely the bufferRow, at least because I need it to calculate the relative row number, and it might be needed in the view anyway.
Where should I put the displayRow recalculation ? Ideally, with normal line numbers, recalculation should only be triggered when adding or removing lines in the editor, not on scroll nor on cursor move. So this update cannot be in updateLinesState, but it cannot be in updateLineNumbersState either, because updateLineNumbersState should only happen on content change...

**Comment by nathansobo** [comment]
@Nadrieril You'll probably need a pretty extensive change to the presenter when in relative line number mode that updates line numbers when the cursor moves instead of based on content.

**Comment by Nadrieril** [comment]
@nathansobo Yeah, I guessed... That's why I haven't I haven't progressed a lot since

**Comment by Francescu** [comment]
Any updates? This feature would be amazing.

**Comment by Nadrieril** [comment]
Atom has changed too much since I made this feature. I would have to code it from scratch, but I don't have time. If anyone is interested, feel free to look at my code

**Pull request closed without merging**